lib/hackmode/index.js

```diff
@@ -503,69 +581,227 @@ class N3hHackMode extends N3hMode {
+        // data should be an entryWithMeta
+        // log.t(this.me() + '  dataHoldRequest: ', e.data)
+        const data = JSON.parse(Buffer.from(e.data, 'base64'))
+        // Check if we are receiving only meta
+        const hasEntry = !(Object.entries(data.entry).length === 0 && data.entry.constructor === Object)
+        log.t(this.me() + ' dataHoldRequest:', hasEntry, data)
+        // Store data in Mem & data DHT (done by indexers
+        let fromAgentId = data.dataFetchProviderAgentId
+        if (hasEntry) {
+          fromAgentId = data.entry.providerAgentId
+          this._getMemRef(data.dnaAddress).insert(data.entry)
+        }
+        for (const metaItem of data.meta) {
+          this._getMemRef(data.dnaAddress).insertMeta({
+            type: 'dhtMeta',
+            providerAgentId: fromAgentId,
+            entryAddress: data.entryAddress,
+            attribute: metaItem.attribute,
+            contentList: metaItem.contentList
+          })
+        }
+        break
+      case 'dataFetch':
+        // Issued by _dht._fetchDataLocal()
+        // My dht asked me for all the data of an entry, give it a response
+        log.t(this.me() + ' dataFetch', e)
+        // Get entryAddress
+        const entryAddress = e.dataAddress
+        // log.e(this.me() + ' entryAddress', entryAddress)
+
+        // Search for the DNA with that entryAddress
+        // TODO: bug prone if many dnas hold the same entry ; _handleDhtEvent should provide dnaAddress
+        let mem = null
+        let dna = null
+        for (const dnaAddress in this._memory) {
+          if (this._memory[dnaAddress].has(entryAddress)) {
+            mem = this._memory[dnaAddress]
+            dna = dnaAddress
+            // log.t(this.me() + ' found dataFetch DNA ' + dnaAddress)
+            break
+          }
+        }
+        // If data not found, respond with null
+        if (!mem) {
+          log.e(this.me() + ' doest not have entry ' + entryAddress)
+          const nullB64 = Buffer.from(JSON.stringify(null), 'utf8').toString('base64')
+          this._dht.post(Dht.DhtEvent.dataFetchResponse(e.msgId, nullB64))
+          break
+        }
+        // Create DHT data item
+        let dhtData = mem.get(entryAddress)
+        dhtData.entryAddress = entryAddress
+        dhtData.dnaAddress = dna
+        dhtData.dataFetchProviderAgentId = null
+        // Pick first agent who tracks this DNA as the provider
+        // WARNING non-deterministic if many agents track the same DNA
+        for (const [agentId, dnas] of this._ipcDnaByAgent) {
+          if (dnas.has(dna)) {
+            dhtData.dataFetchProviderAgentId = agentId
+            break
+          }
+        }
+        if (!dhtData.dataFetchProviderAgentId) {
+          log.e(this.me() + ' AgentId not found for DNA ' + dna + ' ; during handleDhtEvent(dataFetch)')
+        }
+        log.t(this.me() + ' dataFetchResponse:', dhtData)
```

Comment:
am I missing the part where we actually ask core for the data, or is that not implemented yet?

Reply:
Mem is 'caching' everything Core sent us (during `publishEntry` and `handleGetPublishingListResult`), so we actually don't need to ask Core for data.
Unless we want to fetch data that has not been published?

Reply:
This is cool for now. Eventually we don't want to duplicate the storage that core is doing.

Reply:
👍